lib/install/deps.js

```diff
@@ -613,8 +635,12 @@ var earliestInstallable = exports.earliestInstallable = function (requiredBy, tr
   // if this tree location requested the same module then we KNOW it
```

Comment:
Ok, so let's back up and talk about what `earliestInstallable` is supposed to do:

You give it a tree, the node in the tree with this requirement and the package.json data for the package and it tells you the location closest to the root of the tree that you can install the package in without producing conflicts. If it can't find any such location then it just returns the requiring node.

I think the problem here is that this comment is a bit off base. We do make one assumption:

* You already know that you need to install this module, you just don't know where yet.

Ordinarily this code is pointless given the existence of `_phantomChildren`– but `_phantomChildren` isn't populated in two circumstances:

1. If you requested the installation of just one (or a few) named modules, eg `npm install thingy`.
2. If you asked for `--only=dev` on a folder without prod deps installed.
3. Or relatedly, if you did an ordinary install and we still have a separate dev deps stage.

So having written this out it seems clear to me that it should be doing a version match like `findRequirement` does to see if the version we have in our hands is compatible with the spec in dependencies (and devDependencies) and if it isn't returning null.

I think that would look something like:
```js
var deps = tree.package.dependencies || {}
if (deps[pkg.name] && !doesPkgVersionMatch(deps[pkg.name], pkg)) return null
var devDeps = tree.package.devDependencies || {}
if (devDeps[pkg.name] && !doesPkgVersionMatch(devDeps[pkg.name], pkg)) return null
```

With a `doesPkgVersionMatch` being split from `doesChildVersionMatch` like:
```js
function doesChildVersionMatch (child, requested, requestor) {
  // we always consider deps provided by a shrinkwrap as "correct" or else
  // we'll subvert them if they're intentionally "invalid"
  if (child.parent === requestor && child.fromShrinkwrap) return true
  return doesPkgVersionMatch(requested, child.package)
}

function doesPkgVersionMatch (requested, pkg) {
  // ranges of * ALWAYS count as a match, because when downloading we allow
  // prereleases to match * if there are ONLY prereleases
  if (requested.spec === '*') return true

  var pkgReq = pkg._requested
  if (pkgReq) {
    if (pkgReq.rawSpec === requested.rawSpec) return true
    if (pkgReq.type === requested.type && pkgReq.spec === requested.spec) return true
  }
  if (!registryTypes[requested.type]) return requested.rawSpec === pkg._from
  return semver.satisfies(child.package.version, requested.spec)
}
```